src/main/java/de/fearnixx/jeak/teamspeak/voice/connection/VoiceConnection.java

```diff
@@ -0,0 +1,223 @@
+import de.fearnixx.jeak.voice.connection.IVoiceConnectionInformation;
+import de.fearnixx.jeak.voice.sound.AudioType;
+import de.fearnixx.jeak.voice.sound.IAudioPlayer;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+public class VoiceConnection implements IVoiceConnection {
+
+    private final AbstractVoiceConnectionInformation clientConnectionInformation;
+    private final String hostname;
+    private final int port;
+    private final IEventService eventService;
+
+    private LocalTeamspeakClientSocket ts3jClientSocket;
+    private boolean locked;
+
+    private boolean connected;
+
+    private IUserService userService;
+
+    private IBot bot;
+
+    private boolean shouldForwardTextMessages;
+
+    VoiceConnection(AbstractVoiceConnectionInformation clientConnectionInformation, String hostname, int port, IEventService eventService, IBot bot, IUserService userService) {
+        this.clientConnectionInformation = clientConnectionInformation;
+        this.hostname = hostname;
+        this.port = port;
+        this.eventService = eventService;
+        this.userService = userService;
+        this.bot = bot;
+        locked = true;
+    }
+
+    @Override
+    public void connect(Runnable onSuccess, Runnable onError) {
+        if (connected) {
+            return;
+        }
+
+        new Thread(
+                () -> {
+                    this.ts3jClientSocket = new LocalTeamspeakClientSocket();
+
+                    ts3jClientSocket.setNickname(clientConnectionInformation.getClientNickname());
+                    ts3jClientSocket.setIdentity(clientConnectionInformation.getTeamspeakIdentity());
+
+                    ts3jClientSocket.addListener(
+                            new TS3Listener() {
+                                @Override
+                                public void onTextMessage(TextMessageEvent e) {
+                                    eventService.fireEvent(
+                                            new VoiceConnectionTextMessageEvent(clientConnectionInformation.getIdentifier(), e)
+                                    );
+
+                                    if (shouldForwardTextMessages) {
+                                        final QueryEvent.ClientTextMessage clientTextMessage = new QueryEvent.ClientTextMessage(userService);
+
+                                        final int invokerId = e.getInvokerId();
+                                        final IClient client = userService.getClientByID(invokerId).orElseThrow();
+
+                                        clientTextMessage.setClient(client);
+                                        clientTextMessage.setCaption(EventCaptions.TEXT_MESSAGE);
+                                        clientTextMessage.setConnection(bot.getServer().getConnection());
```

Comment:
``#getConnection`` may return ``null``. Please cancel/swallow the event in case it is.

Reply:
It seems like a NoSuchElementException would be thrown in this case.